### PR TITLE
feat: add pre commit hook that checks if make proto-all should be run

### PR DESCRIPTION
**Pull request by PanGan21**
## Description

Adds pre commit hook that checks if `make proto-all` should be run.

closes: #6668 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [x] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

File: scripts/hooks/pre-commit.sh [hunk 1]
```diff
@@ -30,5 +30,24 @@ function lint_and_add_modified_go_files() {
...
+function is_proto_all_required() {
+  local before_files=$(git status --porcelain | awk '{print $2}')
+  make proto-all
+  local after_files=$(git status --porcelain | awk '{print $2}')
+  local changed_files=$(comm -13 <(echo "$before_files" | sort) <(echo "$after_files" | sort))
+  if [[ -n "$changed_files" ]]; then
+    echo "Error: Please run 'make proto-all' and commit the updated files."
+    
+    # Revert the changes made by make proto-all
+    for file in $changed_files; do
```
Comment (PanGan21) [inline scripts/hooks/pre-commit.sh:42]: A drawback I can see with this approach is that the pre commit hook becomes quite slower.
I was thinking of the approach to just check if any `.proto` file is changed but this doesn't guarantee whether the command should be run

File: scripts/hooks/pre-commit.sh [hunk 2]
```diff
@@ -30,5 +30,24 @@ function lint_and_add_modified_go_files() {
   done
 }
 
+function is_proto_all_required() {
+  local before_files=$(git status --porcelain | awk '{print $2}')
+  make proto-all
+  local after_files=$(git status --porcelain | awk '{print $2}')
+  local changed_files=$(comm -13 <(echo "$before_files" | sort) <(echo "$after_files" | sort))
+  if [[ -n "$changed_files" ]]; then
+    echo "Error: Please run 'make proto-all' and commit the updated files."
```
[the remark this answers is not in the document]
Comment (bznein) [reply scripts/hooks/pre-commit.sh:39]: That was my idea, yeah. This is still better than what we have now, but having it ran automatically would be ideal to me

[the remark this answers is not in the document]
Comment (PanGan21) [reply scripts/hooks/pre-commit.sh]: Thanks for the review!
I changed the script to run `make proto-all` and do `git add` to the generated files. Le me know what you think :)

File: scripts/hooks/pre-commit.sh (hunk 1 above)
Comment (PanGan21) [reply scripts/hooks/pre-commit.sh]: Refactored. Now the script checks in the unstaged .proto files. If it finds a change it runs `make proto-all` and logs the files that were changed. If it doesn't find a change in protos then it just exits

**Comment by bznein** [comment]
hi @PanGan21, unfortunately we realized that a pre-commit would be very time-consuming and disrupting for the dev workflow and we decided to abandon this issue. Thanks for taking to time to work on it, really appreciated!

**Pull request closed without merging**